packages/kit/src/runtime/server/endpoint.js

```diff
@@ -29,16 +30,29 @@ export async function render_endpoint(event, route) {
 		return method_not_allowed(mod, method);
 	}
 
-	const response = await handler(
-		/** @type {import('types').RequestEvent<Record<string, any>>} */ (event)
-	);
-
-	if (!(response instanceof Response)) {
-		return new Response(
-			`Invalid response from route ${event.url.pathname}: handler should return a Response object`,
-			{ status: 500 }
+	try {
+		const response = await handler(
+			/** @type {import('types').RequestEvent<Record<string, any>>} */ (event)
 		);
-	}
 
-	return response;
+		if (!(response instanceof Response)) {
+			return new Response(
+				`Invalid response from route ${event.url.pathname}: handler should return a Response object`,
+				{ status: 500 }
+			);
+		}
+
+		return response;
+	} catch (error) {
+		if (error instanceof HttpError) {
+			return new Response(error.message, { status: error.status });
+		} else if (error instanceof Redirect) {
+			return new Response(undefined, {
+				status: error.status,
+				headers: { Location: error.location }
+			});
+		} else {
+			throw error;
```

Comment:
Should this be converted into a `Response`, too? If it isn't this means unexpected errors get a regular "page" treatment and they are rendered within possible `+error.svelte` pages which seems wrong.

Reply:
I think so, yeah

Reply:
though actually... i wonder if that should happen at the bottom of `respond`. We already have a `TODO` comment there asking this exact question.

that way we get all the `options.handle_error` and sanitization logic centralised

Reply:
gah, ok that's not possible because we want to render an error page if user code errors in `handle` (at least, that's the decision we made at some point in the past).

i do feel like there's some opportunities for simplification around error handling generally. might noodle in a separate PR. will merge this in the meantime though, no point in delaying it